pkg/build/strategies/dockerfile/dockerfile.go

```diff
@@ -0,0 +1,355 @@
+		return nil, s2ierr.NewUserNotAllowedError(config.AssembleUser, false)
+	}
+
+	dir, _ := filepath.Split(config.AsDockerfile)
+	config.PreserveWorkingDir = true
+	config.WorkingDir = dir
+
+	if config.BuilderImage == "" {
+		builder.result.BuildInfo.FailureReason = utilstatus.NewFailureReason(
+			utilstatus.ReasonGenericS2IBuildFailed,
+			utilstatus.ReasonMessageGenericS2iBuildFailed,
+		)
+		return builder.result, errors.New("builder image name cannot be empty")
+	}
+
+	if err := builder.Prepare(config); err != nil {
+		return builder.result, err
+	}
+
+	if err := builder.CreateDockerfile(config); err != nil {
+		builder.result.BuildInfo.FailureReason = utilstatus.NewFailureReason(
+			utilstatus.ReasonDockerfileCreateFailed,
+			utilstatus.ReasonMessageDockerfileCreateFailed,
+		)
+		return builder.result, err
+	}
+
+	builder.result.Success = true
+
+	return builder.result, nil
+}
+
+// CreateDockerfile takes the various inputs and creates the Dockerfile used by
+// the docker cmd to create the image produced by s2i.
+func (builder *Dockerfile) CreateDockerfile(config *api.Config) error {
+	glog.V(4).Infof("Constructing image build context directory at %s", config.WorkingDir)
+	buffer := bytes.Buffer{}
+
+	if len(config.ImageWorkDir) == 0 {
+		config.ImageWorkDir = "/opt/app-root/src"
+	}
+
+	imageUser := config.AssembleUser
+
+	// where files will land inside the new image.
+	scriptsDestDir := filepath.Join(getDestination(config), "scripts")
+	sourceDestDir := filepath.Join(getDestination(config), "src")
+
+	buffer.WriteString(fmt.Sprintf("FROM %s\n", config.BuilderImage))
+
+	if len(config.AssembleUser) > 0 {
+		buffer.WriteString(fmt.Sprintf("USER %s\n", imageUser))
+	}
+
+	generatedLabels := util.GenerateOutputImageLabels(builder.sourceInfo, config)
+	if len(generatedLabels) > 0 || len(config.Labels) > 0 {
+		first := true
+		buffer.WriteString("LABEL ")
+		for k, v := range generatedLabels {
+			if !first {
+				buffer.WriteString(fmt.Sprintf(" \\\n      "))
+			}
+			buffer.WriteString(fmt.Sprintf("%q=%q", k, v))
+			first = false
+		}
+		for k, v := range config.Labels {
+			if !first {
+				buffer.WriteString(fmt.Sprintf(" \\\n      "))
+			}
+			buffer.WriteString(fmt.Sprintf("%q=%q", k, v))
+			first = false
+		}
+
+		buffer.WriteString("\n")
+	}
+
+	// only COPY scripts dir if required scripts are present, i.e. the dir is not empty;
+	// even if the "scripts" dir exists, the COPY would fail if it was empty
+	scriptsProvided, fileNames := checkValidDirWithContents(filepath.Join(config.WorkingDir, builder.uploadScriptsDir))
+	assembleProvided := false
+	runProvided := false
+	for _, f := range fileNames {
+		glog.V(2).Infof("found override script file %s", f.Name())
+		if f.Name() == "run" {
+			runProvided = true
+		} else if f.Name() == "assemble" {
+			assembleProvided = true
+		}
+		if runProvided && assembleProvided {
+			break
+		}
+	}
+
+	env := createBuildEnvironment(config.WorkingDir, config.Environment)
```

Comment:
Does the build env need to be generated after the scripts check is run? Based on the inputs it seems the ENV directives can be written prior to the scripts check.

Reply:
yeah it can be done either way.  i'll move the env logic up before the script check.